src/lib/efhw/af_xdp.c

```diff
@@ -151,6 +151,17 @@ struct event_waiter
   int budget;
 };
 
+/* Ring memory mapping artifacts */
+struct ring_map {
+  int n_pages;
+  struct page** pages;
+  void* vmapped_addr;
+};
+
+/* Are rings allocated as physically-continuous memory (linux<6.3) or
+ * via vmalloc (linux>=6.3)? */
+static bool rings_are_physically_continuous = true;
```

Comment:
I'm not sure we need this variable. I think it would be simpler to call `follow_pfn` for each ring unconditionally, and see whether it succeeds.

Reply:
I disagree with you.

These 2 cases are completely different.  It is lucky that (for now) we can determine the case by `follow_pfn`.  I'm not sure this hack will work forever.  And yes, I believe it is a hack.  So, I believe we need some sentinel which case is running here.  I'd prefer compile-time condition, but it is not easy.  Such a run-time variable is better than nothing.

Reply:
Yes, that's my idea.